lib/mongo/socket.rb

```diff
@@ -123,6 +123,7 @@ def initialize(family)
     def read(length)
       handle_errors do
         data = read_from_socket(length)
+        raise IOError unless data.length == length
```

Comment:
The docs are wrong in that case then and should be changed... As for the failures... Probably to do with the mocking... I can have a look as well.